test_plans/SYCL-2020/aspect_selector.md

```diff
@@ -0,0 +1,102 @@
+# Test plan for the aspect selector
+
+This is the test plan for the `aspect_selector` feature of SYCL 2020.
+
+## Testing scope
+
+### Device coverage
+
+All tests construct a test device (i.e. `testDevice`),
+representing the device where conformance will be passed on.
+Each test will construct an aspect selector object,
+which will try to select another device,
+this one based on the requested aspects (i.e. `selectedDevice`).
+`selectedDevice` can be any device in the system,
+not necessarily the same as `testDevice`.
+
+### Function overloads
+
+Each test that tests aspects `{asp1, ..., aspN}` must test all three overloads of the `aspect_selector` function:
+
+* Pass a vector of aspects: `aspect_selector(std::vector{asp1, ..., aspN})`
+* Pass aspects as function arguments: `aspect_selector(asp1, ..., aspN)`
+* Pass aspects as template parameters: `aspect_selector<asp1, ..., aspN>()`
+
+### Combinations of aspects
+
+All aspects listed in the `enum class aspect` will be tested,
+let's identify these aspects as `{asp1, ..., aspN}`.
+
+The CTS will cover these combinations:
+
+1. All combinations where no aspect is passed in.
+2. All combinations where a single aspect is used.
+3. All combinations of two aspects,
+including duplicates (e.g. `aspect_selector(aspX, aspX)`),
+and reversed order (e.g. `aspect_selector(aspX, aspY)` and `aspect_selector(aspY, aspX)`).
+
+The CTS does not have to test all combinations of 3 or more aspects.
+
+The CTS will test all the different numbers of aspects at least once, up to `N`,
+where the aspects are not duplicated,
+i.e. `aspect_selector(asp1), aspect_selector(asp1, asp2), ..., aspect_selector(asp1, asp2, ..., aspN)`.
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
```

Comment:
We had an internal discussion recently about adding tests with randomized inputs to the CTS, and we thought it was not a good idea.  We agree that such tests are useful, but we're not sure they belong in the CTS.  Vendors use the CTS to demonstrate conformance, so it seems potentially unfair for the CTS to behave differently for one vendor vs. another.  Consider two vendors that have the same bug.  One might run the CTS and get a pass while another runs the CTS and gets a fail.  That seems unfair.

Reply:
On the other hand we can imagine a test-driven implementation of SYCL implementing *only* the CTS... ;-)

Reply:
How about requiring that the generated tests must always produce the same combinations (by using a predefined seed)? I've updated that in the latest wording.

Reply:
Yes, I think this is reasonable.  As long at the test runs the same for each vendor, I'm happy.